File: scripts/ui_audit_remote.py
from pathlib import Path

from playwright.sync_api import sync_playwright


URL = "https://driftline-xvxczqg62a-uc.a.run.app/"
OUT = Path("/tmp/driftline-ui-audit")
OUT.mkdir(parents=True, exist_ok=True)


def audit_view(page, name: str) -> dict[str, object]:
    page.screenshot(path=str(OUT / f"{name}.png"), full_page=True)
    metrics = page.evaluate(
        """() => ({
          viewportWidth: window.innerWidth,
          documentWidth: document.documentElement.scrollWidth,
          bodyWidth: document.body.scrollWidth,
          horizontalOverflow: document.documentElement.scrollWidth > window.innerWidth + 1,
          headings: [...document.querySelectorAll('h1,h2')].map((node) => node.textContent.trim()).slice(0, 30),
          buttons: [...document.querySelectorAll('button')].map((node) => node.textContent.trim()).filter(Boolean).slice(0, 30),
          dialogs: document.querySelectorAll('[role="dialog"]').length,
        })"""
    )
    return metrics


with sync_playwright() as playwright:
    browser = playwright.chromium.launch(headless=True)
    page = browser.new_page(viewport={"width": 1440, "height": 1000}, device_scale_factor=1)
    console_errors: list[str] = []
    page_errors: list[str] = []
    page.on("console", lambda message: console_errors.append(message.text) if message.type == "error" else None)
    page.on("pageerror", lambda error: page_errors.append(str(error)))
    page.goto(URL, wait_until="domcontentloaded", timeout=60_000)
    page.wait_for_load_state("networkidle", timeout=60_000)
    page.wait_for_timeout(1500)
    desktop = audit_view(page, "desktop-initial")
    page.get_by_role("button", name="View source evidence").first.click()
    page.get_by_role("dialog").wait_for()
    evidence_open = audit_view(page, "desktop-evidence")
    page.get_by_role("button", name="Close source evidence").click()

    run = page.get_by_role("button", name="Run scan")
    run.click()
    page.get_by_text("Scan complete", exact=False).wait_for(timeout=180_000)
    page.wait_for_timeout(500)
    workflow = audit_view(page, "desktop-pending-approval")
    page.get_by_role("button", name="Open evidence").first.click()
    page.get_by_role("dialog").wait_for()
    pending_evidence = audit_view(page, "desktop-pending-evidence")
    page.get_by_role("button", name="Close source evidence").click()

    page.set_viewport_size({"width": 390, "height": 844})
    page.wait_for_timeout(500)
    mobile = audit_view(page, "mobile-pending-approval")
    print({
        "desktop": desktop,
        "evidence_open": evidence_open,
        "workflow": workflow,
        "pending_evidence": pending_evidence,
        "mobile": mobile,
        "console_errors": console_errors,
        "page_errors": page_errors,
    })
    browser.close()

File: scripts/ui_audit_remote.mjs
import { mkdir } from "node:fs/promises";
import { chromium } from "/var/folders/5x/sfc5z0cn04g36dpqyzd4sghc0000gn/T/tmp.Knj1dS52MO/node_modules/playwright/index.mjs";

const url = "https://driftline-xvxczqg62a-uc.a.run.app/";
const out = "/tmp/driftline-ui-audit";
await mkdir(out, { recursive: true });
const browser = await chromium.launch({
  headless: true,
  executablePath: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
});
const page = await browser.newPage({ viewport: { width: 1440, height: 1000 }, deviceScaleFactor: 1 });
const consoleErrors = [];
const pageErrors = [];
page.on("console", (message) => { if (message.type() === "error") consoleErrors.push(message.text()); });
page.on("pageerror", (error) => pageErrors.push(String(error)));
await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60_000 });
await page.waitForLoadState("networkidle", { timeout: 60_000 });
await page.waitForTimeout(1500);
const view = async (name) => {
  await page.screenshot({ path: `${out}/${name}.png`, fullPage: true });
  return page.evaluate(() => ({
    viewportWidth: window.innerWidth,
    documentWidth: document.documentElement.scrollWidth,
    bodyWidth: document.body.scrollWidth,
    horizontalOverflow: document.documentElement.scrollWidth > window.innerWidth + 1,
    headings: [...document.querySelectorAll("h1,h2")].map((node) => node.textContent.trim()).slice(0, 30),
    buttons: [...document.querySelectorAll("button")].map((node) => node.textContent.trim()).filter(Boolean).slice(0, 30),
    dialogs: document.querySelectorAll('[role="dialog"]').length,
  }));
};
const desktop = await view("desktop-initial");
await page.getByRole("button", { name: "View source evidence" }).first().click();
await page.getByRole("dialog").waitFor();
const evidenceOpen = await view("desktop-evidence");
await page.getByRole("button", { name: "Close source evidence" }).click();
await page.getByRole("button", { name: "Run scan" }).click();
await page.getByText("Scan complete", { exact: false }).waitFor({ timeout: 180_000 });
await page.waitForTimeout(500);
const workflow = await view("desktop-pending-approval");
await page.getByRole("button", { name: "Open evidence" }).first().click();
await page.getByRole("dialog").waitFor();
const pendingEvidence = await view("desktop-pending-evidence");
await page.getByRole("button", { name: "Close source evidence" }).click();
await page.setViewportSize({ width: 390, height: 844 });
await page.waitForTimeout(500);
const mobile = await view("mobile-pending-approval");
console.log(JSON.stringify({ desktop, evidenceOpen, workflow, pendingEvidence, mobile, consoleErrors, pageErrors }, null, 2));
await browser.close();
